### PR TITLE
Add evaluation util / tool for quickly comparing perf across backends [Part 1]

### PR DESCRIPTION
**Pull request by voznesenskym**
I wrote this for myself while onboarding an OpenAI model, cleaned it up so others can use it.

# Run 1

**Command**

```
torchdynamo.evaluate(
        model, backends=[None, "inductor", "aot_cudagraphs"], inputs=images, file=f
    )
```

**Output**

```
Starting evaluation of 3 backends: {'inductor', None, 'aot_cudagraphs'}
EVALUATING inductor
EXPLAIN FOR inductor: Dynamo produced 1 graphswith 0 graph break and 4 ops
 Break reasons: 

TorchDynamo compilation metrics:
Function                                               Runtimes (s)
---------------------------------------------------  --------------
convert_frame_assert.<locals>._convert_frame_assert          0.2435
WARMUP FINISHED FOR inductor
         2308095 function calls (2193061 primitive calls) in 12.637 seconds

   Ordered by: internal time
   List reduced from 3745 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      327    2.815    0.009    2.815    0.009 {built-in method io.open_code}
     1267    2.605    0.002    2.605    0.002 {built-in method posix.stat}
      340    2.187    0.006    2.187    0.006 {method '__exit__' of '_io._IOBase' objects}
     4000    1.491    0.000    1.491    0.000 {built-in method torch._ops.aten.convolution}
     1000    0.773    0.001    3.101    0.003 /tmp/torchinductor_voz/hk/chkefwghhp2gjolndfqgusvyqepzysx7zhcrm23tibmkufagv7tw.py:271(call)
        5    0.727    0.145    0.727    0.145 {built-in method posix.waitpid}
       35    0.240    0.007    0.240    0.007 {built-in method posix.listdir}
        2    0.085    0.042    0.085    0.043 /data/home/voz/miniconda/envs/torchdynamo3.10/lib/python3.10/site-packages/triton/ops/matmul.py:12(get_configs_io_bound)
     8020    0.075    0.000    0.075    0.000 {built-in method torch.empty_strided}
     1000    0.060    0.000    3.188    0.003 {built-in method apply}


RUN FINISHED FOR inductor
    RUN FOR inductor TOOK: 13.398822784423828 seconds for 1000 inputs and 0 warmups
ALL RESULTS EQUAL ACROSS ALL BACKENDS
EVALUATING No Dynamo
EXPLAIN FOR No Dynamo: Dynamo produced 1 graphswith 0 graph break and 4 ops
 Break reasons: 

TorchDynamo compilation metrics:
Function                                               Runtimes (s)
---------------------------------------------------  --------------
convert_frame_assert.<locals>._convert_frame_assert          0.0406
WARMUP FINISHED FOR No Dynamo
         147010 function calls (137010 primitive calls) in 2.456 seconds

   Ordered by: internal time
   List reduced from 33 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     4000    1.266    0.000    1.266    0.000 {built-in method torch.conv2d}
     4000    0.367    0.000    0.367    0.000 {built-in method torch.relu}
     4000    0.334    0.000    0.334    0.000 {built-in method torch.batch_norm}
     4000    0.105    0.000    1.384    0.000 /scratch/voz/work/pytorch/torch/nn/modules/conv.py:462(forward)
     1000    0.047    0.000    0.047    0.000 {built-in method torch.max_pool2d}
     4000    0.045    0.000    0.045    0.000 {method 'add_' of 'torch._C._TensorBase' objects}
    44000    0.040    0.000    0.040    0.000 /scratch/voz/work/pytorch/torch/nn/modules/module.py:1252(__getattr__)
     4000    0.040    0.000    0.498    0.000 /scratch/voz/work/pytorch/torch/nn/modules/batchnorm.py:137(forward)
     1000    0.036    0.000    2.452    0.002 /scratch/voz/work/torchdynamo/t7.py:56(forward)
     1000    0.035    0.000    0.035    0.000 {built-in method torch._C._nn.linear}


RUN FINISHED FOR No Dynamo
    RUN FOR No Dynamo TOOK: 2.4581568241119385 seconds for 1000 inputs and 0 warmups
ALL RESULTS EQUAL ACROSS ALL BACKENDS
EVALUATING aot_cudagraphs
EXPLAIN FOR aot_cudagraphs: Dynamo produced 1 graphswith 0 graph break and 4 ops
 Break reasons: 

TorchDynamo compilation metrics:
Function                                               Runtimes (s)
---------------------------------------------------  --------------
convert_frame_assert.<locals>._convert_frame_assert          0.0382
WARMUP FINISHED FOR aot_cudagraphs
         961734 function calls (908770 primitive calls) in 0.762 seconds

   Ordered by: internal time
   List reduced from 1267 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     1000    0.117    0.000    2.495    0.002 {built-in method apply}
     1000    0.055    0.000    2.330    0.002 <eval_with_key>.25:4(forward)
23431/4123    0.036    0.000    0.118    0.000 /scratch/voz/work/pytorch/torch/utils/_pytree.py:139(tree_flatten)
22299/9125    0.023    0.000    0.044    0.000 /scratch/voz/work/pytorch/torch/fx/node.py:603(map_aggregate)
    23431    0.019    0.000    0.027    0.000 /scratch/voz/work/pytorch/torch/utils/_pytree.py:112(__init__)
162261/158816    0.019    0.000    0.022    0.000 {built-in method builtins.isinstance}
3004/1000    0.018    0.000    3.100    0.003 /scratch/voz/work/torchdynamo/torchdynamo/eval_frame.py:163(_fn)
17088/16586    0.018    0.000    2.556    0.000 /scratch/voz/work/pytorch/torch/_ops.py:253(__call__)
1326/1079    0.013    0.000    0.247    0.000 /scratch/voz/work/pytorch/torch/_subclasses/fake_tensor.py:660(__torch_dispatch__)
    34104    0.013    0.000    0.016    0.000 /scratch/voz/work/pytorch/torch/utils/_pytree.py:86(_is_namedtuple_instance)


RUN FINISHED FOR aot_cudagraphs
    RUN FOR aot_cudagraphs TOOK: 3.105978488922119 seconds for 1000 inputs and 0 warmups
ALL RESULTS EQUAL ACROSS ALL BACKENDS
```

# Run 2

**Command**
```
with open("output.txt", "a") as f:
    torchdynamo.evaluate(
        model, backends=[None, "inductor", "aot_cudagraphs"], warmups=images[0:10], inputs=images[11:90], file=f
    )

```
**Output**

```
Starting evaluation of 3 backends: {'aot_cudagraphs', None, 'inductor'}
EVALUATING aot_cudagraphs
EXPLAIN FOR aot_cudagraphs: Dynamo produced 1 graphswith 0 graph break and 4 ops
 Break reasons: 

TorchDynamo compilation metrics:
Function                                               Runtimes (s)
---------------------------------------------------  --------------
convert_frame_assert.<locals>._convert_frame_assert          0.1813
WARMUP FINISHED FOR aot_cudagraphs
         5376 function calls (5218 primitive calls) in 0.018 seconds

   Ordered by: internal time
   List reduced from 40 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       79    0.004    0.000    0.154    0.002 {built-in method apply}
       79    0.004    0.000    0.147    0.002 <eval_with_key>.10:4(forward)
       79    0.003    0.000    0.003    0.000 {built-in method torch._ops.aten.}
   237/79    0.001    0.000    0.156    0.002 /scratch/voz/work/torchdynamo/torchdynamo/eval_frame.py:164(_fn)
     1264    0.001    0.000    0.143    0.000 /scratch/voz/work/pytorch/torch/_ops.py:253(__call__)
       79    0.000    0.000    0.149    0.002 /scratch/voz/work/pytorch/functorch/_src/aot_autograd.py:414(forward)
       79    0.000    0.000    0.149    0.002 /scratch/voz/work/pytorch/functorch/_src/aot_autograd.py:246(call_func_with_args)
       79    0.000    0.000    0.000    0.000 /scratch/voz/work/pytorch/torch/nn/modules/module.py:1252(__getattr__)
       79    0.000    0.000    0.147    0.002 /scratch/voz/work/pytorch/torch/fx/graph_module.py:262(__call__)
       79    0.000    0.000    0.147    0.002 /scratch/voz/work/pytorch/torch/nn/modules/module.py:1184(_call_impl)


RUN FINISHED FOR aot_cudagraphs
    RUN FOR aot_cudagraphs TOOK: 0.15629291534423828 seconds for 79 inputs and 10 warmups
ALL RESULTS EQUAL ACROSS ALL BACKENDS
EVALUATING No Dynamo
EXPLAIN FOR No Dynamo: Dynamo produced 1 graphswith 0 graph break and 4 ops
 Break reasons: 

TorchDynamo compilation metrics:
Function                                               Runtimes (s)
---------------------------------------------------  --------------
convert_frame_assert.<locals>._convert_frame_assert          0.0454
WARMUP FINISHED FOR No Dynamo
         11617 function calls (10827 primitive calls) in 0.215 seconds

   Ordered by: internal time
   List reduced from 31 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      316    0.133    0.000    0.133    0.000 {built-in method torch.conv2d}
      316    0.029    0.000    0.029    0.000 {built-in method torch.relu}
      316    0.026    0.000    0.026    0.000 {built-in method torch.batch_norm}
      316    0.003    0.000    0.003    0.000 {method 'add_' of 'torch._C._TensorBase' objects}
       79    0.003    0.000    0.003    0.000 {built-in method torch.max_pool2d}
      316    0.003    0.000    0.037    0.000 /scratch/voz/work/pytorch/torch/nn/modules/batchnorm.py:137(forward)
     3476    0.003    0.000    0.003    0.000 /scratch/voz/work/pytorch/torch/nn/modules/module.py:1252(__getattr__)
       79    0.003    0.000    0.003    0.000 {built-in method torch._C._nn.linear}
       79    0.003    0.000    0.215    0.003 /scratch/voz/work/torchdynamo/t7.py:75(forward)
   869/79    0.002    0.000    0.215    0.003 /scratch/voz/work/pytorch/torch/nn/modules/module.py:1184(_call_impl)


RUN FINISHED FOR No Dynamo
    RUN FOR No Dynamo TOOK: 0.21541166305541992 seconds for 79 inputs and 10 warmups
ALL RESULTS EQUAL ACROSS ALL BACKENDS
EVALUATING inductor
EXPLAIN FOR inductor: Dynamo produced 1 graphswith 0 graph break and 4 ops
 Break reasons: 

TorchDynamo compilation metrics:
Function                                               Runtimes (s)
---------------------------------------------------  --------------
convert_frame_assert.<locals>._convert_frame_assert          0.0386
WARMUP FINISHED FOR inductor
         10906 function calls (10669 primitive calls) in 0.268 seconds

   Ordered by: internal time
   List reduced from 40 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      316    0.108    0.000    0.108    0.000 {built-in method torch._ops.aten.convolution}
       79    0.073    0.001    0.255    0.003 /tmp/torchinductor_voz/hk/chkefwghhp2gjolndfqgusvyqepzysx7zhcrm23tibmkufagv7tw.py:271(call)
      316    0.055    0.000    0.055    0.000 {built-in method torch._ops.aten.}
       79    0.010    0.000    0.268    0.003 {built-in method apply}
      632    0.008    0.000    0.008    0.000 {built-in method torch.empty_strided}
      711    0.004    0.000    0.004    0.000 {built-in method torch.as_strided}
   237/79    0.002    0.000    0.271    0.003 /scratch/voz/work/torchdynamo/torchdynamo/eval_frame.py:164(_fn)
     1580    0.002    0.000    0.002    0.000 {method 'size' of 'torch._C._TensorBase' objects}
     1264    0.001    0.000    0.001    0.000 {method 'stride' of 'torch._C._TensorBase' objects}
     1896    0.001    0.000    0.001    0.000 {method 'data_ptr' of 'torch._C._TensorBase' objects}


RUN FINISHED FOR inductor
    RUN FOR inductor TOOK: 0.2713460922241211 seconds for 79 inputs and 10 warmups
ALL RESULTS EQUAL ACROSS ALL BACKENDS
```

### CODE REVIEW

File: torchdynamo/eval_frame.py [hunk 1]
```diff
@@ -364,6 +369,105 @@ def toy_example(a, b):
...
+            backend_fn = torchdynamo.optimize(backend, nopython=nopython)(f)
+
+        first_input = inputs[0]
+        (
+            explanation,
+            _,
+            _,
+            _,
+            _,
+        ) = torchdynamo.explain(f, first_input)
```
Comment (NouamaneTazi) [inline torchdynamo/eval_frame.py:418]: Current implementation doesn't seem to work for functions with multiple args. We can fix that by forcing `inputs` to be of type list of tuples with the following fix
```suggestion
        ) = torchdynamo.explain(f, *first_input)
```

File: torchdynamo/eval_frame.py [hunk 2]
```diff
@@ -364,6 +369,105 @@ def toy_example(a, b):
...
+            raise e
+
+        try:
+            pr = cProfile.Profile()
+            pr.enable()
+
+            start = time.time()
+            for input_idx in range(0, len(inputs)):
+                input = inputs[input_idx]
+                result = backend_fn(input)
```
Comment (NouamaneTazi) [inline torchdynamo/eval_frame.py:438]: Same as previous comment
```suggestion
                result = backend_fn(*input)
```

**Review by NouamaneTazi — CHANGES_REQUESTED** [review]

File: torchdynamo/eval_frame.py (hunk 1 above)
Comment (voznesenskym) [reply torchdynamo/eval_frame.py:418]: Ah yeah, good find. I should add some tests to this ;)

File: torchdynamo/eval_frame.py (hunk 2 above)
Comment (voznesenskym) [reply torchdynamo/eval_frame.py:438]: Same as previous comment

File: torchdynamo/eval_frame.py [hunk 3]
```diff
@@ -364,6 +369,105 @@ def toy_example(a, b):
     )
 
 
+def evaluate(
```
Comment (jansel) [inline torchdynamo/eval_frame.py:372]: I fell like this should live in torchdynamo.testing

**Review by jansel — CHANGES_REQUESTED** [review]

File: torchdynamo/eval_frame.py (hunk 3 above)
Comment (voznesenskym) [reply torchdynamo/eval_frame.py:372]: This isn't used in tests. Why there? Maybe a new dir torchdynamo.tooling?

Comment (jansel) [reply torchdynamo/eval_frame.py:372]: `torchdynamo.testing` has other tools for measuring perf and checking results.  It is not unit test related.

**Review by jansel — APPROVED** [review]

**Comment by jansel** [comment]
We have migrated `torchdynamo` to `torch._dynamo` and will use the `pytorch/pytorch` repo for future development.  Please resubmit this PR to https://github.com/pytorch/pytorch/

More details and instructions to port this PR over can be found in https://github.com/pytorch/torchdynamo/issues/1588

**Pull request closed without merging**